[WGSL] Add a name mangling pass

tadeuzagallo opened the pull request:
#### 4697eaccd00f7f8e11b01e96fe00b5976d8e047f
<pre>
[WGSL] Add a name mangling pass
<a href="https://bugs.webkit.org/show_bug.cgi?id=251441">https://bugs.webkit.org/show_bug.cgi?id=251441</a>
rdar://104872078

Reviewed by Myles C. Maxfield.

Initial name mangling pass implementation. It&apos;s still a bit bare bones, and until
we implement the type checking we can&apos;t rename struct fields, but everything else
(that is currently supported) works. The implementation is a bit rough, just
renaming everything in place, but it&apos;s fairly concise and gets the job done, at
least for the time being.

* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTStructureAccess.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/CallGraph.h:
(WGSL::CallGraph::ast const):
(WGSL::CallGraph::entrypoints const):
(WGSL::CallGraph::ast): Deleted.
(WGSL::CallGraph::entrypoints): Deleted.
* Source/WebGPU/WGSL/MangleNames.cpp: Added.
(WGSL::MangledName::toString const):
(WGSL::NameManglerVisitor::NameManglerVisitor):
(WGSL::NameManglerVisitor::Context::Context):
(WGSL::NameManglerVisitor::Context::lookup const):
(WGSL::NameManglerVisitor::Context::add):
(WGSL::NameManglerVisitor::ContextScope::ContextScope):
(WGSL::NameManglerVisitor::ContextScope::~ContextScope):
(WGSL::NameManglerVisitor::run):
(WGSL::NameManglerVisitor::visit):
(WGSL::NameManglerVisitor::visitFunctionBody):
(WGSL::NameManglerVisitor::visitVariableDeclaration):
(WGSL::NameManglerVisitor::def):
(WGSL::NameManglerVisitor::makeMangledName):
(WGSL::NameManglerVisitor::read const):
(WGSL::mangleNames):
* Source/WebGPU/WGSL/MangleNames.h: Copied from Source/WebGPU/WGSL/CallGraph.h.
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::getNamedFunction const):

Canonical link: <a href="https://commits.webkit.org/259671@main">https://commits.webkit.org/259671@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40430baa892277a982367bedbd41fc7414223147

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105638 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14729 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38513 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114866 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5928 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111392 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12270 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95249 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/97904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26888 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7997 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/8495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14114 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47789 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6681 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10046 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/ShaderModule.mm:
```diff
@@ -236,9 +236,12 @@ auto compilationMessageData(convertMessages(
+id<MTLFunction> ShaderModule::getNamedFunction(const String& originalName, const HashMap<String, double>& keyValueReplacements) const
 {
+    const auto* information = entryPointInformation(originalName);
+    const String& name = information ? information->mangledName : originalName;
```
oh nice :)

mwyrzykowski reviewed (approved):
Looks great, will need approval from a WebKit reviewer

litherum commented on Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
```diff
@@ -57,7 +57,7 @@ class Parameter final : public Node {
     }
 
     Kind kind() const override;
-    const String& name() const { return m_name; }
```
The _getter_ can continue to be const, right?

litherum commented on Source/WebGPU/WGSL/WGSL.cpp:
```diff
@@ -99,17 +100,26 @@ PrepareResult prepare(AST::ShaderModule& ast, const HashMap<String, PipelineLayo
+        RUN_PASS_WITH_RESULT(entryPointMap, mangleNames, callGraph);
+
+        for (const auto& it : entryPointMap) {
+            Reflection::EntryPointInformation information;
```
Oh, cool, I'm assuming you're doing this to report the mangled entry point names to the API?

litherum commented on Source/WebGPU/WGSL/MangleNames.cpp:
```diff
@@ -0,0 +1,273 @@
+
+void NameManglerVisitor::visitVariableDeclaration(AST::VariableDecl& variable, MangledName::Kind kind)
+{
+    def(variable.name(), kind);
```
Can we use full words instead of `def`?

litherum commented on Source/WebGPU/WGSL/MangleNames.cpp:
```diff
@@ -0,0 +1,273 @@
+
+void NameManglerVisitor::visit(AST::IdentifierExpression& identifier)
+{
+    read(identifier.identifier());
```
Not super obvious what `read` means

tadeuzagallo committed:
[WGSL] Add a name mangling pass
https://bugs.webkit.org/show_bug.cgi?id=251441
rdar://104872078

Reviewed by Myles C. Maxfield.

Initial name mangling pass implementation. It's still a bit bare bones, and until
we implement the type checking we can't rename struct fields, but everything else
(that is currently supported) works. The implementation is a bit rough, just
renaming everything in place, but it's fairly concise and gets the job done, at
least for the time being.

* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTIdentifierExpression.h:
* Source/WebGPU/WGSL/AST/ASTStructureAccess.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/CallGraph.h:
(WGSL::CallGraph::ast const):
(WGSL::CallGraph::entrypoints const):
(WGSL::CallGraph::ast): Deleted.
(WGSL::CallGraph::entrypoints): Deleted.
* Source/WebGPU/WGSL/MangleNames.cpp: Added.
(WGSL::MangledName::toString const):
(WGSL::NameManglerVisitor::NameManglerVisitor):
(WGSL::NameManglerVisitor::Context::Context):
(WGSL::NameManglerVisitor::Context::lookup const):
(WGSL::NameManglerVisitor::Context::add):
(WGSL::NameManglerVisitor::ContextScope::ContextScope):
(WGSL::NameManglerVisitor::ContextScope::~ContextScope):
(WGSL::NameManglerVisitor::run):
(WGSL::NameManglerVisitor::visit):
(WGSL::NameManglerVisitor::visitFunctionBody):
(WGSL::NameManglerVisitor::visitVariableDeclaration):
(WGSL::NameManglerVisitor::def):
(WGSL::NameManglerVisitor::makeMangledName):
(WGSL::NameManglerVisitor::read const):
(WGSL::mangleNames):
* Source/WebGPU/WGSL/MangleNames.h: Copied from Source/WebGPU/WGSL/CallGraph.h.
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/ShaderModule.mm:
(WebGPU::ShaderModule::getNamedFunction const):

Canonical link: https://commits.webkit.org/259671@main
